scrunch/order.py

```diff
@@ -0,0 +1,576 @@
+            return
+
+        _elements = collections.OrderedDict()
+        for item in items:
+            _elements[item] = self.elements[item]
+        self.elements = _elements
+
+        self.order.update()
+
+    def create_group(self, name, alias=None, position=-1, before=None,
+                     after=None):
+        name = self._validate_name_arg(name)
+        # when we want to create an empty group
+        if not alias:
+            self.elements[name] = Group(
+                {name: []}, order=self.order, parent=self)
+            self.order.update()
+            return
+        elements = self._validate_alias_arg(alias)
+        position = 0 if (before or after) else position
+
+        # create the new Group obj and insert all `elements`
+        new_group = Group({name: []}, order=self.order, parent=self)
+        new_group.insert(elements)
+
+        # add the new Group to self.elements so that `insert` detects it
+        self.elements[name] = new_group
+        self.insert(new_group.name, position=position,
+                    before=before, after=after)
+
+    def rename(self, name):
+        name = self._validate_name_arg(name)
+
+        if self.is_root:
+            raise ValueError(
+                'Renaming the root Group is not allowed.'
+            )
+
+        if name in self.parent.elements:
+            raise ValueError(
+                'Parent Group \'%s\' already contains an element named \'%s\'.'
+                % (self.parent.name, name)
+            )
+
+        if name == self.name:
+            # Nothing to do.
+            return
+
+        # Rename!
+        _elements = collections.OrderedDict()
+        for current_name, obj in self.parent.elements.items():
+            if current_name == self.name:
+                _elements[name] = obj
+            else:
+                _elements[current_name] = obj
+        self.name = name
+        self.parent.elements = _elements
+
+        # Update!
+        self.order.update()
+
+    def move(self, path, position=-1, before=None, after=None):
+        position = 0 if (before or after) else position
+        path = Path(path)
+        if not path.is_absolute:
+            raise InvalidPathError(
+                'Invalid path %s: only absolute paths are allowed.' % path
+            )
+        target_group = self.order[str(path)]
+
+        if target_group == self:
+            raise InvalidPathError(
+                'Invalid path %s: cannot move Group into itself.' % path
+            )
+        target_group.insert(
+            self.name, position=position, before=before, after=after
+        )
+
+
+class Order(object):
+
+    def __init__(self, resource):
+        self.resource = resource
+        self._hier = None
+        self._graph = None
+        self._sync = True
+
+    def _load_hier(self):
+        """
+        subclasses need to overwrite this method in order to get the updated
+        order hierarchy
+        """
+        return self._hier
```

Comment:
Looks like _hier maps to the shoji:order resource? Why it needs a separate getter? You already receive the resource on init.